scripts/release.ts

```diff
@@ -0,0 +1,229 @@
+import { build } from './build';
+
+// TODO: Change to import when this is converted to TS
+const { uiLink, uiLine } = require('../lib/ui');
+
+const exec = promisify(_exec);
+
+const MAIN_BRANCH = 'main';
+
+const TAG = {
+  LATEST: 'latest',
+  NEXT: 'next',
+  EXPERIMENTAL: 'experimental',
+} as const;
+
+const INCREMENT = {
+  PATCH: 'patch',
+  MINOR: 'minor',
+  MAJOR: 'major',
+  PRERELEASE: 'prerelease',
+} as const;
+
+const VERSION_INCREMENT_OPTIONS = [
+  INCREMENT.PATCH,
+  INCREMENT.MINOR,
+  INCREMENT.MAJOR,
+  INCREMENT.PRERELEASE,
+] as const;
+const TAG_OPTIONS = [TAG.LATEST, TAG.NEXT, TAG.EXPERIMENTAL] as const;
+
+type ReleaseArguments = {
+  versionIncrement: typeof VERSION_INCREMENT_OPTIONS[number];
+  tag: typeof TAG_OPTIONS[number];
+};
+
+type DistTags = {
+  [TAG.LATEST]: string;
+  [TAG.NEXT]: string;
+  [TAG.EXPERIMENTAL]: string;
+};
+
+type Tag = typeof TAG_OPTIONS[number];
+
+async function getGitBranch(): Promise<string> {
+  const { stdout } = await exec('git rev-parse --abbrev-ref HEAD');
+  return stdout.trim();
+}
+
+async function getDistTags(): Promise<DistTags> {
+  const { stdout } = await exec(`npm view ${packageName} dist-tags --json`);
+  const distTags = stdout.trim();
+  return JSON.parse(distTags) as DistTags;
+}
+
+async function cleanup(newVersion: string): Promise<void> {
+  await exec(`git reset HEAD~`);
+  await exec(`git checkout .`);
+  await exec(`git tag -d v${newVersion}`);
+}
+
+async function publish(tag: Tag): Promise<void> {
+  logger.log();
+  logger.log(`Publishing to ${tag}...`);
+  uiLine();
+  logger.log();
+
+  return new Promise((resolve, reject) => {
+    const childProcess = spawn('npm', ['publish', '--dry-run', '--tag', tag], {
+      stdio: 'inherit',
+      cwd: './dist',
+    });
+
+    childProcess.on('close', code => {
+      if (code !== EXIT_CODES.SUCCESS) {
+        reject();
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+async function handler({
+  versionIncrement,
+  tag,
+}: ArgumentsCamelCase<ReleaseArguments>): Promise<void> {
+  setLogLevel(LOG_LEVEL.LOG);
+
+  const branch = await getGitBranch();
+
+  const isExperimental = tag === TAG.EXPERIMENTAL;
+
+  if (isExperimental && branch === MAIN_BRANCH) {
+    logger.error(
+      'Releases to experimental tag cannot be published from the main branch'
+    );
+    process.exit(EXIT_CODES.ERROR);
+  } else if (!isExperimental && branch !== MAIN_BRANCH) {
+    logger.error(
+      'Releases to latest and next tags can only be published from the main branch'
+    );
+    process.exit(EXIT_CODES.ERROR);
+  }
+
+  if (tag === TAG.LATEST && versionIncrement === INCREMENT.PRERELEASE) {
+    logger.error(
+      'Invalid release: cannot increment prerelease number on latest tag.'
+    );
+  }
+
+  const {
+    next: currentNextTag,
+    experimental: currentExperimentalTag,
+  } = await getDistTags();
+
+  if (!isExperimental && currentNextTag !== localVersion) {
```

Comment:
My suggestion [here](https://github.com/HubSpot/hubspot-cli/pull/1228/files#r1829439062) would probably make more sense to implement here (rather than where I commented it).

Just to enumerate the potential scenarios

**User is releasing to the experimental tag:** N/A nothing to prevent here

**User is releasing to next tag:** if local version contains a next tag, it should be equivalent to the `currentNextTag`. But it's possible that the local version won't contain one (i.e. the last publish was a publish to latest). In that case, the local version should have the same base as the `currentNextTag`.

**User is releasing to latest tag:** Check to see if local version is equal to `currentNextTag` AND that the `currentLatestTag` + `versionIncrement` would be equivalent to the base of the local version. This means that the local changes have already been released to the next tag, and can safely be released to latest. If not, we should prompt b/c it's dangerous to release changes directly to latest.